File: sdk/go/examples/simple_consumer.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/rustmq/rustmq/sdk/go/rustmq"
)

func main() {
	// Create client configuration
	config := rustmq.DefaultClientConfig()
	config.Brokers = []string{"localhost:9092"}
	config.ClientID = "simple-consumer-go"

	// Create client
	client, err := rustmq.NewClient(config)
	if err != nil {
		log.Fatalf("Failed to create client: %v", err)
	}
	defer client.Close()

	fmt.Println("Connected to RustMQ cluster")

	// Create consumer
	consumer, err := client.CreateConsumer("example-topic", "example-group")
	if err != nil {
		log.Fatalf("Failed to create consumer: %v", err)
	}
	defer consumer.Close()

	fmt.Println("Created consumer for topic: example-topic, group: example-group")

	// Handle Ctrl+C gracefully
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)

	messageCount := 0
	done := make(chan bool)

	// Start message consumption in a goroutine
	go func() {
		defer func() { done <- true }()

		for {
			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
			
			message, err := consumer.Receive(ctx)
			cancel()

			if err != nil {
				if err == context.DeadlineExceeded {
					// Timeout is normal, continue
					continue
				}
				if err.Error() == "consumer is closed" {
					fmt.Println("Consumer closed, stopping message processing")
					break
				}
				fmt.Printf("Error receiving message: %v\n", err)
				time.Sleep(1 * time.Second)
				continue
			}

			if message == nil {
				fmt.Println("Consumer closed")
				break
			}

			messageCount++
			
			fmt.Printf("Received message #%d: id=%s, offset=%d, partition=%d\n", 
				messageCount, message.Message.ID, message.Message.Offset, message.Partition())
			
			fmt.Printf("  Payload: %s\n", message.Message.PayloadAsString())
			
			// Print headers
			for key, value := range message.Message.Headers {
				fmt.Printf("  Header %s: %s\n", key, value)
			}
			
			// Acknowledge the message
			if err := message.Ack(); err != nil {
				fmt.Printf("Failed to acknowledge message: %v\n", err)
			} else {
				fmt.Println("  Message acknowledged")
			}
			
			// Show message age
			fmt.Printf("  Message age: %v\n", message.Message.Age())
		}
	}()

	// Wait for shutdown signal
	<-sigChan
	fmt.Println("\nReceived shutdown signal, closing consumer...")

	// Close consumer (this will break the receive loop)
	if err := consumer.Close(); err != nil {
		fmt.Printf("Error closing consumer: %v\n", err)
	}

	// Wait for message processing to finish
	select {
	case <-done:
		fmt.Println("Message processing stopped")
	case <-time.After(5 * time.Second):
		fmt.Println("Timeout waiting for message processing to stop")
	}

	// Show consumer metrics
	metrics := consumer.Metrics()
	fmt.Printf("Consumer metrics: received=%d, processed=%d, failed=%d, lag=%d\n", 
		metrics.MessagesReceived, metrics.MessagesProcessed, 
		metrics.MessagesFailed, metrics.Lag)

	fmt.Printf("Consumer and client closed. Total messages processed: %d\n", messageCount)
}

File: sdk/rust/examples/simple_consumer.rs
use rustmq_client::*;
use tokio;
use tracing_subscriber;
use tokio::signal;

#[tokio::main]
async fn main() -> Result<()> {
    // Initialize logging
    tracing_subscriber::fmt::init();

    // Create client configuration
    let config = ClientConfig {
        brokers: vec!["localhost:9092".to_string()],
        client_id: Some("simple-consumer".to_string()),
        ..Default::default()
    };

    // Create client
    let client = RustMqClient::new(config).await?;
    println!("Connected to RustMQ cluster");

    // Create consumer
    let consumer = client.create_consumer("example-topic", "example-group").await?;
    println!("Created consumer for topic: example-topic, group: example-group");

    // Handle Ctrl+C gracefully
    let consumer_clone = consumer.clone();
    tokio::spawn(async move {
        signal::ctrl_c().await.expect("Failed to install Ctrl+C handler");
        println!("\nReceived Ctrl+C, shutting down...");
        consumer_clone.close().await.ok();
    });

    // Consume messages
    let mut message_count = 0;
    loop {
        match consumer.receive().await {
            Ok(Some(consumer_message)) => {
                message_count += 1;
                
                let message = &consumer_message.message;
                println!("Received message #{}: id={}, offset={}, partition={}", 
                    message_count, message.id, message.offset, message.partition);
                
                if let Ok(payload) = message.payload_as_string() {
                    println!("  Payload: {}", payload);
                }
                
                // Print headers
                for (key, value) in &message.headers {
                    println!("  Header {}: {}", key, value);
                }
                
                // Acknowledge the message
                if let Err(e) = consumer_message.ack().await {
                    eprintln!("Failed to acknowledge message: {}", e);
                }
                
                println!("  Message acknowledged");
            }
            Ok(None) => {
                println!("Consumer closed");
                break;
            }
            Err(e) => {
                eprintln!("Error receiving message: {}", e);
                tokio::time::sleep(tokio::time::Duration::from_millis(1000)).await;
            }
        }
    }

    // Close client
    client.close().await?;
    println!("Consumer and client closed. Total messages processed: {}", message_count);

    Ok(())
}